naga/tests/out/glsl/subgroup-operations-s.main.Compute.glsl

```diff
@@ -0,0 +1,58 @@
+#version 430 core
+#extension GL_ARB_compute_shader : require
+#extension GL_KHR_shader_subgroup_basic : require
+#extension GL_KHR_shader_subgroup_vote : require
+#extension GL_KHR_shader_subgroup_arithmetic : require
+#extension GL_KHR_shader_subgroup_ballot : require
+#extension GL_KHR_shader_subgroup_shuffle : require
+#extension GL_KHR_shader_subgroup_shuffle_relative : require
+layout(local_size_x = 1, local_size_y = 1, local_size_z = 1) in;
+
+uint num_subgroups_1 = 0u;
+
+uint subgroup_id_1 = 0u;
+
+uint subgroup_size_1 = 0u;
+
+uint subgroup_invocation_id_1 = 0u;
+
+
+void main_1() {
+    uint _e5 = subgroup_size_1;
+    uint _e6 = subgroup_invocation_id_1;
+    uvec4 _e9 = subgroupBallot(((_e6 & 1u) == 1u));
+    uvec4 _e10 = subgroupBallot(true);
+    bool _e12 = subgroupAll((_e6 != 0u));
+    bool _e14 = subgroupAny((_e6 == 0u));
+    uint _e15 = subgroupAdd(_e6);
+    uint _e16 = subgroupMul(_e6);
+    uint _e17 = subgroupMin(_e6);
+    uint _e18 = subgroupMax(_e6);
+    uint _e19 = subgroupAnd(_e6);
+    uint _e20 = subgroupOr(_e6);
+    uint _e21 = subgroupXor(_e6);
+    uint _e22 = subgroupExclusiveAdd(_e6);
+    uint _e23 = subgroupExclusiveMul(_e6);
+    uint _e24 = subgroupInclusiveAdd(_e6);
+    uint _e25 = subgroupInclusiveMul(_e6);
+    uint _e26 = subgroupBroadcastFirst(_e6);
+    uint _e27 = subgroupBroadcast(_e6, 4u);
+    uint _e30 = subgroupShuffle(_e6, ((_e5 - 1u) - _e6));
+    uint _e31 = subgroupShuffleDown(_e6, 1u);
+    uint _e32 = subgroupShuffleUp(_e6, 1u);
+    uint _e34 = subgroupShuffleXor(_e6, (_e5 - 1u));
+    return;
+}
+
+void main() {
+    uint num_subgroups = gl_NumSubgroups;
+    uint subgroup_id = gl_SubgroupID;
+    uint subgroup_size = gl_SubgroupSize;
```

Comment:
This mapping is potentially incorrect unless the VK_EXT_subgroup_size_control extension is enabled and the VK_PIPELINE_SHADER_STAGE_CREATE_ALLOW_VARYING_SUBGROUP_SIZE_BIT_EXT flag is set. Both are on by default in Vulkan 1.3, but I'm pretty sure this code is intended to run on pre-1.3 versions.

I personally would recommend enabling the extension and turning on the flag on instance creation, and only advertising the subgroup extension when it is available. Failing that, there are workarounds which might have problems, such as dividing the workgroup size by the number of subgroups; this has been tested to work on Intel GPUs without the size control extension turned on.

Reply:
Can you elaborate how the mapping is incorrect / not what is expected?

I understand subgroup size to mean the number of possible lanes, not the (potentially smaller) number of active lanes. That can instead be obtained by doing popcount on a subgroup ballot.

Also, as mentioned in the other discussion, so far I only experienced two scenarios is which a compute shader / kernel is launched with partially deactivated subgroups:
1. The grid size is not divisible by the subgroup size
2. The workgroup size is not divisible by the subgroup size

Other than these I never saw an implementation do actually heterogeneous subgroup sizes (fully activated but different groupings) in the same kernel launch.

Reply:
I wrote about this a bit in my [Prefix sum on Vulkan](https://raphlinus.github.io/gpu/2020/04/30/prefix-sum.html#subgroups-and-subgroup-size) blog post, and in followup issues.

My personal take on this is that modeling the subgroup size as always 32 on Intel, yet masking lanes so only 8 or 16 is active, is a polite fiction to work around a design limitation of Vulkan 1.1, fixed as of Vulkan 1.3, and should be rejected as a mistake, undesirable to ever expose to a WebGPU client. Writing code that is defensive against all possible subgroup masks is basically impossible. Doing popcount on a ballot is definitely another possible workaround for determining `subgroup_size`, and may be a better approach to the if subgroup size control is not available, but I think there's a risk it can't be recognized as a compile time constant.

Reply:
Lets say the workgroup size is 15 and the hardware is running at a subgroup size of 8. But subgroups are partially deactivated from the start because of register pressure. The workitem in workgroup indices are given in hex or `#` if inactive.
```
[0 1 2 3 # # # # ]
        [4 5 6 7 # # # # ]
                [8 9 A B # # # # ]
                        [C D E # # # # # ]
```

So what do you expect `subgroup_size` to be in the given example?
1. `8` for all subgroups (which you reject as a mistake)
2. `4` for all subgroups
3. `4` for the first 3 subgroups and `3` for the last (which you could also get from `subgroupBallotBitcount`)

Reply:
I expect 8 for all subgroups, and I don't agree that I classified that as a mistake - I want `subgroup_size` to reflect the hardware subgroup size, and in particular to match the Vulkan 1.3 SubgroupSize semantics.

I'm not sure where your picture comes from, I'm not aware of any case in which Intel will do that. And if the subgroup size is actually 4 (which may be the case for SwiftShader), then I'd expect a subgroup size of 4. Technically it would be allowed behavior with a subgroup size of 8 if the FULL_SUBGROUPS flag is not set, but to me that would be very surprising behavior from any actual GPU.

> If [subgroup size control is enabled] the SubgroupSize decorated variable will contain the subgroup size for each subgroup that gets dispatched. This value must be between [minSubgroupSize](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#limits-minSubgroupSize) and [maxSubgroupSize](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#limits-maxSubgroupSize) and must be uniform with [subgroup scope](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#shaders-scope-subgroup). The value may vary across a single draw call, and for fragment shaders may vary across a single primitive. In compute dispatches, SubgroupSize must be uniform with [command scope](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#shaders-scope-command).

So this clearly rules out your choice 3.

I believe the Vulkan spec provides sufficiently strong guarantees that the popcount of a ballot will match the subgroup size when FULL_SUBGROUPS is enabled. It will, I believe, empirically be true for compute shaders in practice, though I certainly approve of relying on the actual guarantee of the API in cases when the API has taken care to document its guarantees (as Vulkan does).

> Developers can also specify a new pipeline shader stage create flag that requires the implementation to have fully populated subgroups within local workgroups. This requires the workgroup size in the X dimension to be a multiple of the subgroup size.

Validation [02758](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkPipelineShaderStageCreateInfo.html#VUID-VkPipelineShaderStageCreateInfo-flags-02758) further narrows this by saying that the workgroup X dimension must be a multiple of maxSubgroupSize when FULL_SUBGROUPS is set. (Note that this will be enforced for cooperative matrices, so it's worth looking ahead. Also note that life becomes more difficult for the programmer if subgroupMaxSize is not made available.)

Whether the application can request FULL_SUBGROUPS is a TODO in the current [gpuweb proposal](https://github.com/gpuweb/gpuweb/blob/main/proposals/subgroups.md). I would vote yes, because it seems to be needed if you want any kind of guarantee your code will work. The guarantee can reliably be offered on Vulkan; as the proposal states, other APIs are unclear.

There is *strong* value in having the subgroup size potentially available as a compile time constant; the robustly portable code for workgroup-wide prefix sum will have cases for 2, 3, and 4 levels for a workgroup size of 256, and you want those extra branches compiled out. It's an interesting empirical question whether the popcount of the ballot can get constant-folded, perhaps worth testing with microbenchmarks on different GPUs.

Reply:
> I expect 8 for all subgroups, and I don't agree that I classified that as a mistake - I want subgroup_size to reflect the hardware subgroup size, and in particular to match the Vulkan 1.3 SubgroupSize semantics.

Sorry, I might have misinterpreted what you were referring to with:
> should be rejected as a mistake, undesirable to ever expose to a WebGPU client

> I'm not sure where your picture comes from, I'm not aware of any case in which Intel will do that.

The example was not specific to Intel, just a general example to tease out the possible different meanings of "subgroup size".

> SubgroupSize must be uniform with [command scope](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#shaders-scope-command). So this clearly rules out your choice 3.

You are right. But then I don't get what `VK_PIPELINE_SHADER_STAGE_CREATE_ALLOW_VARYING_SUBGROUP_SIZE_BIT` is about, if "SubgroupSize must be uniform with command scope".

> Validation [02758](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkPipelineShaderStageCreateInfo.html#VUID-VkPipelineShaderStageCreateInfo-flags-02758) further narrows this by saying that the workgroup X dimension must be a multiple of maxSubgroupSize when FULL_SUBGROUPS is set. (Note that this will be enforced for cooperative matrices, so it's worth looking ahead. Also note that life becomes more difficult for the programmer if subgroupMaxSize is not made available.)

This does thread into the other discussion above about multi dimensional workgroup sizes and how the logical layout / mapping of workitems to subgroup would work in that case. We have not decided on that yet.

> There is strong value in having the subgroup size potentially available as a compile time constant

Agreed, cascades of subgroup operations are a hassle in general. But I don't see how we could achieve this across all backends.

Reply:
"Varying" is I think confusing naming. For compute shaders, it means that SubgroupSize (gl_SubgroupSize) can be the the actual subgroup size chosen by the shader compiler, as opposed to the previous semantics which forced it to be equal to `subgroupSize` in [VkPhysicalDeviceSubgroupProperties](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkPhysicalDeviceSubgroupProperties.html). That latter value is constant for the physical device, and cannot vary for different shader compilations.

> This does thread into the other discussion above about multi dimensional workgroup sizes and how the logical layout / mapping of workitems to subgroup would work in that case. We have not decided on that yet.

I agree that this is tricky. The Vulkan spec is clear on making no mapping guarantees - In Vulkan, when using subgroups it's best to rely on `subgroup_invocation_id` and use that for buffer indexing, etc, not relying on a specific mapping between that and LocalInvocationId. But then Vulkan provides SubgroupID, which is missing from HLSL. This is a TODO item in the proposal. In practice, for 1D dispatches, I think SubgroupInvocationID + SubgroupID * SubgroupSize will equal LocalInvocationID.x. But also note that this will go horribly wrong if you're on Intel and don't have subgroup size control - SubgroupSize will be 32 and so this will overflow your buffer.

Reply:
Makes sense now, thanks for the explanation!